syntax = "proto3";

package com.grupog.inventario;

option java_multiple_files = true;
option java_package = "com.grupog.inventario";

enum CategoriaInventario {
  ROPA = 0;
  ALIMENTOS = 1;
  JUGUETES = 2;
  UTILES_ESCOLARES = 3;
}

message InventarioItem {
  int64 id = 1;
  CategoriaInventario categoria = 2;
  string descripcion = 3;
  int32 cantidad = 4;
  bool eliminado = 5;
  int64 fechaAlta = 6;
  string usuarioAlta = 7;
  int64 fechaModificacion = 8;
  string usuarioModificacion = 9;
}

message CrearItemRequest {
  CategoriaInventario categoria = 1;
  string descripcion = 2;
  int32 cantidad = 3;
  string usuarioEjecutor = 4;
}

message ActualizarItemRequest {
  int64 id = 1;
  string descripcion = 2;
  int32 cantidad = 3;
  string usuarioEjecutor = 4;
}

message BajaLogicaItemRequest {
  int64 id = 1;
  string usuarioEjecutor = 2;
}

message ListarItemsRequest {
  bool incluirEliminados = 1;
}

message BuscarItemPorIdRequest {
  int64 id = 1;
}

message RegistrarSalidaRequest {
  CategoriaInventario categoria = 1;
  string descripcion = 2;
  int32 cantidad = 3;
  string usuarioEjecutor = 4;
}

message ListaItemsResponse {
  repeated InventarioItem items = 1;
}

message RespuestaExito {
  string mensaje = 1;
  bool exito = 2;
}


// ========== KAFKA - Solicitudes de donaci√≥n ==========
message DonacionItem {
  string categoria = 1;
  string descripcion = 2;
  int32 cantidad = 3;
}

message SolicitarDonacionesRequest {
  repeated DonacionItem donaciones = 1;
}

message SolicitarDonacionesResponse {
  string idSolicitud = 1;
  string mensaje = 2;
  bool exito = 3;
}

// ========== KAFKA - Transferir donaciones ==========
message TransferirDonacionRequest {
  string idSolicitud = 1; // ID de solicitud externa
  repeated DonacionItem donaciones = 2;
}

// ========== Listar solicitudes externas ==========
message ListarSolicitudesExternasRequest {
  bool soloActivas = 1; // true = solo activas, false = todas
}

message SolicitudExterna {
  string idSolicitud = 1;
  int64 idOrganizacionSolicitante = 2;
  repeated DonacionItem donaciones = 3;
  bool activa = 4;
  int64 fechaRecepcion = 5; // timestamp
}

message ListarSolicitudesExternasResponse {
  repeated SolicitudExterna solicitudes = 1;
}


service InventarioService {
  rpc crearItem (CrearItemRequest) returns (InventarioItem);
  rpc actualizarItem (ActualizarItemRequest) returns (InventarioItem);
  rpc bajaLogicaItem (BajaLogicaItemRequest) returns (RespuestaExito);
  rpc listarItems (ListarItemsRequest) returns (ListaItemsResponse);
  rpc buscarItemPorId (BuscarItemPorIdRequest) returns (InventarioItem);
  rpc registrarSalida (RegistrarSalidaRequest) returns (RespuestaExito);
  
  // Kafka - Solicitudes
  rpc solicitarDonaciones (SolicitarDonacionesRequest) returns (SolicitarDonacionesResponse);
  rpc listarSolicitudesExternas (ListarSolicitudesExternasRequest) returns (ListarSolicitudesExternasResponse);
  
  // Kafka - Transferencias
  rpc transferirDonacion (TransferirDonacionRequest) returns (RespuestaExito);
}
